nexus/go-handler.md

```diff
@@ -0,0 +1,307 @@
+			StartWorkflowOptions: StartWorkflowOptions{
+				ID: fmt.Sprintf("%s-%s", request.Operation, input.CellID),
+			},
+		}, nil
+	}))
+```
+
+### Workflow Query
+
+```go
+package operation // go.temporal.io/sdk/operation
+
+type WorkflowQueryOptions struct {
+	WorkflowID           string
+	// Optionally expose RunID too
+	QueryType            string
+	QueryRejectCondition enums.QueryRejectCondition
+	// Input to provide to the query.
+	Input                []any
+	// TODO: Should Temporal Header be settable here?
+}
+
+func NewWorkflowQueryHandler[T any](func(context.Context, T, *StartOperationRequest) (*WorkflowQueryOptions, error)) Handler
+```
+
+**Basic usage:**
+
+```go
+w.RegisterOperation("get-cell-status", NewWorkflowQueryHandler(
+	func(ctx context.Context, input Input, request *StartOperationRequest) (*WorkflowQueryOptions, error) {
+		return &WorkflowQueryOptions{
+			Input:      []any{input},
+			QueryType:  "get-cell-status",
+			WorkflowID: fmt.Sprintf("%s-%s", "provision-cell", input.CellID),
+		}, nil
+	}))
+```
+
+### Workflow Signal
+
+```go
+package operation // go.temporal.io/sdk/operation
+
+type WorkflowSignalOptions struct {
+	WorkflowID           string
+	// Optionally expose RunID too
+	SignalType            string
+	// Input to provide to the query.
+	Input                []any
+	// TODO: Should Temporal Header be settable here?
+}
+
+func NewWorkflowSignalHandler[T any](func(context.Context, T, *StartOperationRequest) (*WorkflowSignalOptions, error)) Handler
+```
+
+**Basic usage:**
+
+```go
+w.RegisterOperation("update-cell-config", NewWorkflowSignalHandler(
+	func(ctx context.Context, input Input, request *StartOperationRequest) (*WorkflowSignalOptions, error) {
+		return &WorkflowSignalOptions{
+			Input:      []any{input},
+			SignalType: "update-cell-config",
+			WorkflowID: fmt.Sprintf("%s-%s", "provision-cell", input.CellID),
+		}, nil
+	}))
+```
+
+### Workflow Update
+
+Not defined yet. We can easily support sync update, but async update would require more thought on the guarantees
+Temporal provides for callback delivery on update completion.
+
+### Multi-tenant handlers
+
+This section introduces a couple of additional concepts, failing handler requests, and operation interceptors.
+
+Handlers should be able to fail requests in a meaningful way, providing the proper HTTP response code to convey errors
+such as bad request and unauthorized. To control request failure, handlers can return a `nexus.HandlerError` with an
+HTTP status code and failure object.
+Returning any other error from a handler should result in an internal server error and the error details omitted from
+the response.
+
+Interceptors are used to share logic across all registered operations. i.e, interceptors can enforce authorized access
+to operations making them ideal for low boilerplate multi-tenant application development.
+
+```go
+package interceptor // go.temporal.io/sdk/interceptor
+
+type OperationInboundInterceptor interface {
```

Comment:
I don't think this is how interceptors should be handled. Handlers are interfaces, so it is very easy to create a wrapper handler that intercepts and does things. That's how most Go middleware works (e.g. how we do data converters, how HTTP does HTTP handler middleware, etc).

Reply:
But the SDK exposes an API to register a handler for an operation, not a generic handler for any operation and there's no concept of chaining handlers.

Do you think that should change?

Reply:
Right, just like Go has the ability to register a handler for an HTTP call. Users are expected to wrap it before registering in these languages. That's how the handler-based middleware works.

> Do you think that should change?

Not necessarily. Anyone can do `RegisterOperation("my-operation", NewAuthHandlerMiddleware(actualOperationHandler))`.

Reply:
Yeah, but they'd have to repeat this wrapping for every operation, I don't like that.

Reply:
But everyone else writing middleware does. If you'd like, you can accept this worker option:

```go
NexusHandlerMiddleware: []func(Handler) Handler
```

Reply:
```
NexusHandlerMiddleware: []func(Handler) Handler
```

That's essentially what the interceptors do, just a different way of exposing this, I think interceptors are more consistent with how we approach middleware in the the other parts of the SDK.

Reply:
At the least you could add them later and not concern yourself so early with the advanced interception feature.

> I think interceptors are more consistent with how we approach middleware in the the other parts of the SDK.

Just know that many people are going to ignore interceptors and do middleware the way they do for other interface-based handlers in Go.